packages/next/next-server/lib/router/router.ts

```diff
@@ -188,6 +188,10 @@ export function resolveHref(
   const base = new URL(currentPath, 'http://n')
   const urlAsString =
     typeof href === 'string' ? href : formatWithValidation(href)
+  // Return because it cannot be routed by the Next.js router
+  if (!isLocalURL(urlAsString)) {
```

Comment:
This check is already further up in the code path:
https://github.com/vercel/next.js/blob/a303ab7e4fc77a6a635216a782233448c1b78a28/packages/next/next-server/lib/router/router.ts#L604

Maybe it makes sense to rather move that check in the `push` and `replace` methods, right before `prepareUrlAs`? That way we short-circuit this case earlier and avoid doing the check twice.

Reply:
Sorry if I'm wrong.
I don't think that implementation would solve the problem of trailing slashes in the href attribute for external links.

Since the return value of resolveHref is the base of the href attribute, I thought I needed to modify the resolveHref.
```
  const { href, as } = React.useMemo(() => {
    const [resolvedHref, resolvedAs] = resolveHref(pathname, props.href, true)
    return {
      href: resolvedHref,
      as: props.as
        ? resolveHref(pathname, props.as)
        : resolvedAs || resolvedHref,
    }
  }, [pathname, props.href, props.as])
```

Reply:
In this version, there will always be two calls to `isLocalURL(url)`, once in `resolveHref` (to skip appending a slash) and once in `change` (to circumvent client side routing). I was just wondering if it was possible to avoid executing this double logic (`isLocalURL` does a url parse on every call). Probably requires too much refactoring to make it worth doing in this PR, so let's leave it.